### ISSUE TITLE
Integrate Turing.jl probablistic programming

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/MLJ.jl#157]
https://github.com/alan-turing-institute/MLJ.jl/issues/47#issue-402035942

I think Turing.jl developer @trappmartin has indicated an interest in this and this will be explored at upcoming Sprint

@tlienart whose is joining the Turing Institute team soon, may also be interested in helping out.

**Comment by trappmartin** [comment]
Correct, I’m looking forward to talk about interfacing Turing and MLJ while I’m in London.

**Comment by fkiraly** [comment]
agreed, it's the right time now.
Suggest we start with vanilla Bayesian linear regression, conjugate priors & Gaussian likelihood.

**Comment by trappmartin** [comment]
Here are a few thought on integrating Turing into MLJ for simple models.

We would need to:

1. Implement a mechanism to store to log pdf values for each observation (training or test) to be able to (a) apply model selection and (b) use a Turing model for predictions. See issue TuringLang/Turing.jl#817

2. Provide an interface to compute the log pdf values, related to [PR 750](https://github.com/TuringLang/Turing.jl/pull/750)

3. Implement a wrapper for Turing. Here is an example implementation that could be used once 1 and 2 are resolved:

```Julia
struct TuringModel
   mf::Function
   algo::Turing.InferenceAlgorithm
end

function MLJBase.fit(m::TuringModel, x)
    mf = m.mf
    algo = m.algo

    model = mf(x)
    return Turing.sample(mf(x), algo)
end

function MLJBase.predict(m::TuringModel, fitresult::MCMCChains.Chains, xnew)
    model = m.mf(xnew)

    vi = Turing.VarInfo()
    model(vi, Turing.SampleFromPrior()) # to obtain VarNames
    vsyms = Turing.RandomVariables.syms(vi)

    y = Vector()
    for s in vsyms
        y = vcat(y, vec( Array(fitresult[1,s,1]) ) )
    end
    
    # This is currently not available and the actual syntax might be different.
    return Turing.logpdf(model, vi, aggregate = false)
end
```

**Comment by vollmersj** [comment]
@trappmartin @tlienart 

One way to stay within more restritive family usch as hierarchial GLMs

1. [https://r-nimble.org/](https://r-nimble.org/)
2. [https://cran.r-project.org/web/packages/rstanarm/index.html](https://cran.r-project.org/web/packages/rstanarm/index.html)
3. [https://cran.r-project.org/web/packages/brms/index.html](https://cran.r-project.org/web/packages/brms/index.html)

We could use one of them as an example. IMHO rstanarm looks as the easiest to adapt.

**Comment by trappmartin** [comment]
I just discussed the integration with people at CBL. I’ll compile an issue with some ideas. Making a list of simple models that we support + example implementations sounds like a good plan!

I’ll link the issue here ASAP.
